src/libcore/iter/adapters/mod.rs

```diff
@@ -1752,6 +1752,95 @@ where
     }
 }
 
+/// An iterator that only accepts elements while `predicate` returns `Some(_)`.
+///
+/// This `struct` is created by the [`map_while`] method on [`Iterator`]. See its
+/// documentation for more.
+///
+/// [`map_while`]: trait.Iterator.html#method.map_while
+/// [`Iterator`]: trait.Iterator.html
+#[must_use = "iterators are lazy and do nothing unless consumed"]
+#[unstable(feature = "iter_map_while", reason = "recently added", issue = "none")]
+#[derive(Clone)]
+pub struct MapWhile<I, P> {
+    iter: I,
+    finished: bool,
+    predicate: P,
+}
+
+impl<I, P> MapWhile<I, P> {
+    pub(super) fn new(iter: I, predicate: P) -> MapWhile<I, P> {
+        MapWhile { iter, finished: false, predicate }
+    }
+}
+
+#[unstable(feature = "iter_map_while", reason = "recently added", issue = "none")]
+impl<I: fmt::Debug, P> fmt::Debug for MapWhile<I, P> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MapWhile").field("iter", &self.iter).field("flag", &self.finished).finish()
+    }
+}
+
+#[unstable(feature = "iter_map_while", reason = "recently added", issue = "none")]
+impl<B, I: Iterator, P> Iterator for MapWhile<I, P>
```

Comment:
`MapWhile` should implement `FusedIterator`

Reply:
Hmmm, I'm not sure why, but `TakeWhile` doesn't implement `DoubleEndedIterator`, are you sure `MapWhile` should implement it?

Reply:
Neither `MapWhile` nor `TakeWhile` can implement `DoubleEndedIterator`, because you can only find the first element that doesn't satisfy the condition from the start.

Reply:
@timvermeulen is it really required that condition must be satisfied from start? Why these examples shouldn't work?
```rust
let vec = vec![Some(0), None, Some(1), Some(2)];
let mut iter = vec.into_iter().map_while(std::convert::identity).rev();

assert_eq!(iter.next(), Some(2));
assert_eq!(iter.next(), Some(1));
assert_eq!(iter.next(), None);
```
```rust
let vec = vec![Some(0), None, Some(1), Some(2)];
let mut iter = vec.into_iter().map_while(std::convert::identity);

assert_eq!(iter.next_back(), Some(2));
assert_eq!(iter.next(), Some(0));
assert_eq!(iter.next_back(), Some(1));
assert_eq!(iter.next(), None);
```

Reply:
Yes, a `DoubleEndedIterator` is required to yield the exact same items back-to-front as it does front-to-back (but obviously in reverse order). This is needed to be able to guarantee that `rev()` only ever reverses the order of the items, rather than yielding items you wouldn't otherwise have gotten.

Reply:
I can't see that requirement nor in `DoubleEndedIterator`'s docs nor in `rev`'s docs. `DoubleEndedIterator`'s docs just says
> An iterator able to yield elements from both ends.

And that's all...

Also, if we'll decide to remove `finished flag`, then this requirement will be satisfied:
```rust
let vec = vec![Some(0), None, Some(1), Some(2)];
let mut iter = vec.into_iter().map_while(std::convert::identity).rev();

assert_eq!(iter.next(), Some(2));
assert_eq!(iter.next(), Some(1));
assert_eq!(iter.next(), None);
assert_eq!(iter.next(), Some(0));
assert_eq!(iter.next(), None);
```
```rust
let vec = vec![Some(0), None, Some(1), Some(2)];
let mut iter = vec.into_iter().map_while(std::convert::identity);

assert_eq!(iter.next(), Some(0));
assert_eq!(iter.next(), None);
assert_eq!(iter.next(), Some(1));
assert_eq!(iter.next(), Some(2));
assert_eq!(iter.next(), None);
```

Reply:
> I can't see that requirement nor in DoubleEndedIterator's docs nor in rev's docs.

I agree that the docs could be a little clearer about this. Nonetheless, I think it goes without saying that it would be highly unexpected if for some generic iterator type `I` calling `rev()` on it would give you completely different items. It wouldn't be very useful, either – you could already call `rev()` before calling `map_while` to achieve the same thing.

I'm not sure what you mean with the first snippet, `MapWhile` doesn't implement `DoubleEndedIterator` so we can't call `rev()` on it in the first place. As for the second snippet, it is perfectly fine behavior to return `Some` after having returned `None` as long as `MapWhile` doesn't implement `FusedIterator`. That's the whole reason `fuse` exists. All that the `finished` flag does is make the iterator fused, which is not the responsibility of an individual iterator.

Reply:
> I'm not sure what you mean with the first snippet, MapWhile doesn't implement DoubleEndedIterator so we can't call rev() on it in the first place. 

With both snippets, I mean that we can (and should) implement `DoubleEndedIterator` for `MapWhile`. It wouldn't "give you completely different items", it would give you the same items in reversed order, as expected.

By the way, [`Take`](https://doc.rust-lang.org/std/iter/struct.Take.html) implements `DoubleEndedIterator` and it could return different results if limit (`n`) < length of the inner iterator.

Reply:
Ah sorry, I misunderstood the point you were making. Things like a `for` loop and `collect` will only call `next` until a `None` is encountered, so those two iterators don't actually produce the same values. The first one produces `2` and `1`, and the second one only produces `0`. What follows after `None` is irrelevant. Therefore, the last element of the (non-reversed) iterator is `1`, and you can only know that by iterating from the start.

`Take` doesn't have this problem because given `n` and the length of the iterator, you can know how many elements to skip from the end to get to the last element (note that `Take` only implements `DoubleEndedIterator` when the inner iterator is an `ExactSizeIterator`, and `Take::next_back` calls `self.iter.nth_back(...)`).

Reply:
Oh, I didn't notice how `DoubleEndedIterator` is implemented for `Take`. Thanks for explanations!